houdini/handlers/games/ninja/water.py

```diff
@@ -1,15 +1,1021 @@
+async def handle_get_game(p: Penguin):
+    """Handle the client entering the game"""
+    seat_id = p.waddle.get_seat_id(p)
+    player = p.waddle.get_player_by_penguin(p)
+
+    # needs to send these or the client dies
+    await p.send_xt('gz')
+    await p.send_xt('jz')
+
+    # CMD_PLAYER_INDEX
+    await p.waddle.send_zm_client(player, "po", seat_id)
+
+    start_game = False
+    if isinstance(p.waddle, CardJitsuWaterLogic):
+        # to wait until all have joined
+        player.joined = True
+        start_game = all(player.joined for player in p.waddle.players)
+    elif isinstance(p.waddle, WaterSenseiLogic):
+        start_game = True
+
+    if start_game:
+        await p.waddle.initiate_board()
+        await p.waddle.initiate_player_cards()
+
+
+@handlers.handler(XTPacket('zm', ext='z'), match=['103'])
+@handlers.waddle(CardJitsuWaterLogic, WaterSenseiLogic)
+async def handle_start_game(p: Penguin):
+    """Handle the client being ready to play"""
+    player = p.waddle.get_player_by_penguin(p)
+    player.ready = True
+
+    if not p.waddle.countdown_task:
+        p.waddle.countdown_task = asyncio.create_task(
+            p.waddle.tick_count_down())
+    if all(map(lambda player: player.ready, p.waddle.players)):
+        p.waddle.timer = 3
+
+
+@handlers.handler(XTPacket('zm', ext='z'), match=['110'])
+@handlers.waddle(CardJitsuWaterLogic, WaterSenseiLogic)
+async def handle_choose_card(p: Penguin, *, card_id: str):
+    """Handle a player clicking on a card (which leaves it on a state where it is ready to use)"""
+    player = p.waddle.get_player_by_penguin(p)
+    await p.waddle.select_card(player, int(card_id))
+
+
+@handlers.handler(XTPacket('zm', ext='z'), match=['120'])
+@handlers.waddle(CardJitsuWaterLogic, WaterSenseiLogic)
+async def handle_player_move(p: Penguin, *, cell_id: str):
+    """Handle a player moving to a different cell/stone"""
+    player = p.waddle.get_player_by_penguin(p)
+    available_cells_by_id = p.waddle.get_playable_cells(player)
+    cell_id = int(cell_id)
+
+    async def send_fail():
+        # CMD_PLAYER_INVALID_THROW
+        await p.waddle.send_zm_client(player, 'pf', f'{player.seat_id}-{cell_id}')
+
+    if cell_id not in available_cells_by_id:
+        return await send_fail()
+
+    cell = available_cells_by_id[cell_id]
+
+    if not cell.can_jump():
+        return await send_fail()
+
+    player.jump(cell)
+
+    # CMD_PLAYER_MOVE
+    await p.waddle.send_zm("pm", f'{player.seat_id}-{cell.uid}')
+
+    last_row = p.waddle.board[-1]
+    row_id = cell.uid // 10
+    # -2 because there are always two off-screen rows
+    if last_row.uid - 2 == row_id:
+        return await p.waddle.gong_game_over(player)
+
+
+@handlers.handler(XTPacket('lz', ext='z'))
+@handlers.waddle(CardJitsuWaterLogic, WaterSenseiLogic)
+async def handle_leave_match(p: Penguin):
+    """Sent by the players if they close the game"""
+
+
+@handlers.handler(XTPacket('zm', ext='z'), match=['121'])
```

Comment:
```suggestion
    await p.waddle.send_zm("pm", f"{player.seat_id}-{cell.uid}")

    last_row = p.waddle.board[-1]
    row_id = cell.uid // 10

    # -2 because there are always two off-screen rows
    if last_row.uid - 2 == row_id:
        return await p.waddle.gong_game_over(player)


@handlers.handler(XTPacket("lz", ext="z"))
@handlers.waddle(CardJitsuWaterLogic, WaterSenseiLogic)
async def handle_leave_match(p: Penguin):
    """Sent by the players if they close the game"""


@handlers.handler(XTPacket("zm", ext="z"), match=["121"])
```